src/java.base/share/classes/java/lang/StableValue.java

```diff
@@ -0,0 +1,746 @@
+     * int function upon being first accessed via the returned int function's
+     * {@linkplain IntFunction#apply(int) apply()} method.
+     * <p>
+     * The provided {@code original} int function is guaranteed to be successfully invoked
+     * at most once per allowed input, even in a multi-threaded environment. Competing
+     * threads invoking the returned int function's
+     * {@linkplain IntFunction#apply(int) apply()} method when a value is already under
+     * computation will block until a value is computed or an exception is thrown by
+     * the computing thread.
+     * <p>
+     * If the provided {@code original} int function throws an exception, it is relayed
+     * to the initial caller and no content is recorded.
+     * <p>
+     * The returned int function is not {@link Serializable}.
+     * <p>
+     * If the provided {@code original} int function recursively calls the returned
+     * int function for the same index, an {@linkplain IllegalStateException} will
+     * be thrown.
+     *
+     * @param size     the size of the allowed inputs in {@code [0, size)}
+     * @param original IntFunction used to compute cached values
+     * @param <R>      the type of results delivered by the returned IntFunction
+     * @throws IllegalArgumentException if the provided {@code size} is negative.
+     */
+    static <R> IntFunction<R> intFunction(int size,
+                                          IntFunction<? extends R> original) {
+        if (size < 0) {
+            throw new IllegalArgumentException();
+        }
+        Objects.requireNonNull(original);
+        return StableIntFunction.of(size, original);
+    }
+
+    /**
+     * {@return a new stable function}
+     * <p>
+     * The returned {@link Function function} is a caching function that, for each allowed
+     * input in the given set of {@code inputs}, records the values of the provided
+     * {@code original} function upon being first accessed via the returned function's
+     * {@linkplain Function#apply(Object) apply()} method.
+     * <p>
+     * The provided {@code original} function is guaranteed to be successfully invoked
+     * at most once per allowed input, even in a multi-threaded environment. Competing
+     * threads invoking the returned function's {@linkplain Function#apply(Object) apply()}
+     * method when a value is already under computation will block until a value is
+     * computed or an exception is thrown by the computing thread.
+     * <p>
+     * If the provided {@code original} function throws an exception, it is relayed to
+     * the initial caller and no content is recorded.
+     * <p>
+     * The returned function is not {@link Serializable}.
+     * <p>
+     * If the provided {@code original} function recursively calls the returned
+     * function for the same input, an {@linkplain IllegalStateException} will
+     * be thrown.
+     *
+     * @param inputs   the set of (non-null) allowed input values
```

Comment:
You might get asked what the behavior is when the inputs set contains the null element.

Reply:
Nice catch. The docs only state that if any _parameter_ is `null`, an NPE is thrown. I've fixed this now.